src/sources/util/tcp.rs

```diff
@@ -118,6 +123,7 @@ pub trait TcpSource: Clone + Send + 'static {
                         let events_in = FramedRead::new(socket, source.decoder())
                             .take_until(tripwire)
                             .filter_map(move |frame| {
+                                shutdown_complete_handle.take();
```

Comment:
I still use .take() here because I'm not sure how else to ensure that the future created here and passed into tokio::spawn a few lines later actually captures the shutdown_complete_handle.  I certainly could call clone() instead, but that's an additional clone where just moving the copy created on line 115 into this future would suffice.

Reply:
So this looks like a stream, might make sense here to write some sort of stream combinator that will hold onto the handle until the stream is completed?